clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

```diff
@@ -1510,8 +1511,24 @@ mlir::Value ScalarExprEmitter::VisitCastExpr(CastExpr *CE) {
     return CGF.getBuilder().createBitcast(CGF.getLoc(E->getSourceRange()), Src,
                                           DstTy);
   }
-  case CK_AddressSpaceConversion:
-    llvm_unreachable("NYI");
+  case CK_AddressSpaceConversion: {
+    Expr::EvalResult Result;
+    if (E->EvaluateAsRValue(Result, CGF.getContext()) &&
+        Result.Val.isNullPointer()) {
+      // If E has side effect, it is emitted even if its final result is a
+      // null pointer. In that case, a DCE pass should be able to
+      // eliminate the useless instructions emitted during translating E.
+      if (Result.HasSideEffects) {
+        llvm_unreachable("NYI");
```

Comment:
None of clang OG test cases hit this line. Possibly dead path, just leave it NYI here.

Reply:
@seven-mile you can chose whether to implement this right away or in a next PR (by leaving it NYI). I suggest some next PR so we can consider this PR done sooner.